File: app/schemas/token_book.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel
from pydantic import ConfigDict
from pydantic import Field

from app.schemas.token_identity import TokenIdentitySummaryResponse


class TokenBookIssueSummaryResponse(BaseModel):

    id: int

    token_identity: TokenIdentitySummaryResponse

    issue_number: int

    status: str

    model_config = ConfigDict(
        from_attributes=True
    )


class TokenBookIssueBase(BaseModel):

    token_identity_id: int = Field(
        ...,
        gt=0
    )

    issue_number: int = Field(
        ...,
        gt=0
    )

    remarks: str | None = Field(
        default=None,
        max_length=255
    )


class TokenBookIssueCreate(TokenBookIssueBase):

    pass


class TokenBookIssueUpdate(BaseModel):

    status: str | None = Field(
        default=None,
        max_length=20
    )

    current_sheet: int | None = Field(
        default=None,
        ge=0
    )

    completion_date: datetime | None = Field(
        default=None
    )

    remarks: str | None = Field(
        default=None,
        max_length=255
    )


class TokenBookIssueResponse(TokenBookIssueBase):

    id: int

    issue_date: datetime

    completion_date: datetime | None

    current_sheet: int

    status: str

    is_active: bool

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


class TokenBookIssueListResponse(BaseModel):

    id: int

    token_identity_id: int

    customer_id: int

    customer_code: str

    customer_name: str

    milk_type_name: str

    token_number: int

    issue_number: int

    issue_date: datetime

    status: str

    current_sheet: int

    is_active: bool

    model_config = ConfigDict(
        from_attributes=True
    )


class TokenBookIssueDetailResponse(BaseModel):

    id: int

    token_identity: TokenIdentitySummaryResponse

    issue_number: int

    issue_date: datetime

    completion_date: datetime | None

    current_sheet: int

    status: str

    remarks: str | None

    is_active: bool

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


class TokenBookPaymentBase(BaseModel):

    token_book_issue_id: int = Field(
        ...,
        gt=0
    )

    payment_mode: str = Field(
        ...,
        max_length=20
    )

    book_price: Decimal = Field(
        ...,
        gt=0
    )

    amount_paid: Decimal = Field(
        default=Decimal("0"),
        ge=0
    )

    remarks: str | None = Field(
        default=None,
        max_length=255
    )


class TokenBookPaymentCreate(TokenBookPaymentBase):

    pass


class TokenBookPaymentUpdate(BaseModel):

    payment_mode: str | None = Field(
        default=None,
        max_length=20
    )

    payment_status: str | None = Field(
        default=None,
        max_length=20
    )

    book_price: Decimal | None = Field(
        default=None,
        gt=0
    )

    amount_paid: Decimal | None = Field(
        default=None,
        ge=0
    )

    remarks: str | None = Field(
        default=None,
        max_length=255
    )


class TokenBookPaymentResponse(TokenBookPaymentBase):

    id: int

    payment_status: str

    balance_amount: Decimal

    payment_date: datetime

    collected_by: int | None

    is_active: bool

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


class TokenBookPaymentListResponse(BaseModel):

    id: int

    token_book_issue_id: int

    customer_id: int

    customer_code: str

    customer_name: str

    payment_mode: str

    payment_status: str

    book_price: Decimal

    amount_paid: Decimal

    balance_amount: Decimal

    payment_date: datetime

    is_active: bool

    model_config = ConfigDict(
        from_attributes=True
    )


class TokenBookPaymentDetailResponse(BaseModel):

    id: int

    token_book_issue: TokenBookIssueSummaryResponse

    payment_mode: str

    payment_status: str

    book_price: Decimal

    amount_paid: Decimal

    balance_amount: Decimal

    payment_date: datetime

    collected_by: int | None

    remarks: str | None

    is_active: bool

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )

File: frontend/src/types/token-book.ts
import type { TokenIdentitySummaryResponse } from "./token-identity";

export type BookIssueStatus = "WAITING" | "ACTIVE" | "COMPLETED";
export type BookPaymentStatus = "PAID" | "PARTIAL" | "PENDING";
export type TokenPaymentMode = "PREPAID" | "POSTPAID";

export interface TokenBookIssueCreate {
  token_identity_id: number;
  issue_number: number;
  remarks?: string | null;
}

export interface TokenBookIssueUpdate {
  status?: BookIssueStatus;
  current_sheet?: number;
  completion_date?: string | null;
  remarks?: string | null;
}

export interface TokenBookIssueListResponse {
  id: number;
  token_identity_id: number;
  customer_id: number;
  customer_code: string;
  customer_name: string;
  milk_type_name: string;
  token_number: number;
  issue_number: number;
  issue_date: string;
  status: string;
  current_sheet: number;
  is_active: boolean;
}

export interface TokenBookIssueDetailResponse {
  id: number;
  token_identity: TokenIdentitySummaryResponse;
  issue_number: number;
  issue_date: string;
  completion_date: string | null;
  current_sheet: number;
  status: string;
  remarks: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface TokenBookIssueResponse {
  id: number;
  token_identity_id: number;
  issue_number: number;
  issue_date: string;
  completion_date: string | null;
  current_sheet: number;
  status: string;
  remarks: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface TokenBookIssueSummaryResponse {
  id: number;
  token_identity: TokenIdentitySummaryResponse;
  issue_number: number;
  status: string;
}

export interface TokenBookPaymentCreate {
  token_book_issue_id: number;
  payment_mode: TokenPaymentMode;
  book_price: number;
  amount_paid: number;
  remarks?: string | null;
}

export interface TokenBookPaymentUpdate {
  payment_mode?: TokenPaymentMode;
  book_price?: number;
  amount_paid?: number;
  remarks?: string | null;
}

export interface TokenBookPaymentListResponse {
  id: number;
  token_book_issue_id: number;
  customer_id: number;
  customer_code: string;
  customer_name: string;
  payment_mode: string;
  payment_status: string;
  book_price: number;
  amount_paid: number;
  balance_amount: number;
  payment_date: string;
  is_active: boolean;
}

export interface TokenBookPaymentDetailResponse {
  id: number;
  token_book_issue: TokenBookIssueSummaryResponse;
  payment_mode: string;
  payment_status: string;
  book_price: number;
  amount_paid: number;
  balance_amount: number;
  payment_date: string;
  collected_by: number | null;
  remarks: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface TokenBookPaymentResponse {
  id: number;
  token_book_issue_id: number;
  payment_mode: string;
  payment_status: string;
  book_price: number;
  amount_paid: number;
  balance_amount: number;
  payment_date: string;
  collected_by: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}
